Try finding commit using short SHA if it is not on HEAD

mikhainin opened the pull request:
# Pull Request Check List

If a dependency specifies a git revision using a short hash, Dulwich does not provide an option to directly address it (see discussion in #9560); however, we can still list revisions and find what we are looking for.

Resolves: #9560

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://python-poetry.org/docs/contributing/) at least once, it will save you unnecessary review cycles! -->

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

<!-- If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing! -->

mikhainin commented:
It seemed to work in my case, but I would appreciate it if someone looked into this and told me if it makes sense.
Happy to add tests - I didn't find them for this functionality but my guess it should be somewhere in `tests/vcs/git`?

schneebuzz commented:
> I didn't find them for this functionality but my guess it should be somewhere in `tests/vcs/git`

Some tests for the `..backend.Git.clone` tests are in `tests/integration/test_utils_vcs_git.py`.

radoering commented:
@jelmer I'd appreciate if you could take a look if it makes sense (from a dulwich point of view).

jelmer commented on src/poetry/vcs/git/backend.py:
```diff
@@ -313,21 +314,47 @@ def _clone(cls, url: str, refspec: GitRefSpec, target: Path) -> Repo:
+                        commit.id,
+                    )
+                    local.refs[b"HEAD"] = commit.id
+                    try:
```
What is this try for? If the object was found in the local repo, why would it be missing here?

mikhainin replied:
Hmm, this is a good question. It probably was me being too cautious. Let me remove it.

jelmer replied:
You've retrieved the commit object, so that will definitely be present. 

One of git's basic invariants (ensured by e.g. fetch, etc) is that if you have the commit object you also have the other objects it refers to*.

[*] This isn't strictly true under some circumstances, e.g. shallows - but that's not relevant here.

radoering commented:
> Happy to add tests - I didn't find them for this functionality but my guess it should be somewhere in `tests/vcs/git`?

We have tests in [tests/integration/test_utils_vcs_git.py](https://github.com/python-poetry/poetry/blob/main/tests/integration/test_utils_vcs_git.py) and in [main/tests/vcs/git/test_backend.py](https://github.com/python-poetry/poetry/blob/main/tests/vcs/git/test_backend.py). Fast tests that do not require an external repository can be put into the latter.

> I can see some integration tests against https://github.com/python-poetry/test-fixture-vcs-repository but every commit there seems to have its own head/tag which does not allow to reproduce the initial issue...

I have no idea if it is feasible, but maybe we can create a local repository in the test?

jelmer commented:
Creating a local repository in e.g. a tempdir should be fast and easy with Dulwich - Repo.init(path) will do what you need.

jelmer commented on src/poetry/vcs/git/backend.py:
```diff
@@ -313,21 +314,47 @@ def _clone(cls, url: str, refspec: GitRefSpec, target: Path) -> Repo:
 
+    @classmethod
+    def _find_object_by_sha_prefix(cls, repo: Repo, sha_prefix: bytes) -> Commit | None:
+        for sha in repo.object_store:
```
With the upcoming release of Dulwich you can also use repo.object_store.iter_prefix, which avoids scanning the entire repository.

mikhainin replied:
Shall we wait till the next release and use the new method?

radoering replied:
I do not think we have to wait unless @jelmer says there will be a release in the next days / few weeks. (It will take at least some weeks until the next Poetry release and I think it makes sense to include this with or without the improvement.)

jelmer replied:
FWIW There's a new release out with this API now.

jelmer replied:
```suggestion
        for sha in repo.object_store.iter_prefix(sha_prefix):
```

mikhainin replied:
I can see that this is implemented in **dulwich** `0.22.5`, which drops support of Python 3.8.
It is a reasonable change, but this pull-request branch still requires compatibility with `3.8`. I'm not sure it must be changed as a part of this PR...

@radoering?

Secrus replied:
We have dropped Python 3.8 on the main branch already, so that should not be an issue

radoering replied:
@mikhainin As Secrus said. Just rebase your branch on main. That is required before merging anyway.

mikhainin replied:
Hmm, this direct replace causes:
```

~/src/poetry/src/poetry/vcs/git/backend.py:342: in _find_object_by_sha_prefix
    for sha in repo.object_store.iter_prefix(sha_prefix):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <DiskObjectStore('/private/var/folders/yq/nvkqzc2s60vdd86cg6_79dgh0000gn/T/pytest-of-mikhailgalanin/pytest-87/popen-gw0/test_short_sha_not_in_head0/test-repo/clone-test/.git/objects')>
prefix = b'895d5f'

    def iter_prefix(self, prefix):
        if len(prefix) < 2:
            yield from super().iter_prefix(prefix)
            return
        seen = set()
        dir = prefix[:2].decode()
        rest = prefix[2:].decode()
>       for name in os.listdir(os.path.join(self.path, dir)):
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/yq/nvkqzc2s60vdd86cg6_79dgh0000gn/T/pytest-of/pytest-87/popen-gw0/test_short_sha_not_in_head0/test-repo/clone-test/.git/objects/89'

.../virtualenvs/poetry-GxhiNo8B-py3.11/lib/python3.11/site-packages/dulwich/object_store.py:1042: FileNotFoundError



```

mikhainin commented:
I added test, I had to introduce a new marker "skip_git_mock" not to mock "git clone" as that one is not ready to mock local clone. 

The rest was copied from `tests/vcs/git/test_system.py`: I moved fixtures into separate files so they can be re-used and added a "middle commit" so the issue was finally reproduced.

jelmer commented:
> I added the test, I had to introduce a new marker "skip_git_mock" not to mock "git clone" as that one is not ready to mock local clone. 
> 
> The rest was copied from `tests/vcs/git/test_system.py`: I moved fixtures into separate files so they can be re-used and added a "middle commit" so the issue was finally reproduced.

LGTM, for what it's worth.

radoering commented on pyproject.toml:
```diff
@@ -193,6 +193,7 @@ addopts = [ "-n", "logical", "-ra", "--strict-config", "--strict-markers" ]
 testpaths = ["tests"]
 markers = [
     "network: mark tests that require internet access",
+    "skip_git_mock: makrs test should not auto-apply git_mock"
```
```suggestion
    "skip_git_mock: mark tests that should not auto-apply git_mock"
```

radoering commented on tests/conftest.py:
```diff
@@ -578,3 +584,49 @@ def project_context(project: str | Path, in_place: bool = False) -> Iterator[Pat
+        no_verify=True,
+    )
+
+    # one commit which is not "head
```
```suggestion
    # one commit which is not "head"
```

radoering commented on tests/conftest.py:
```diff
@@ -578,3 +584,49 @@ def project_context(project: str | Path, in_place: bool = False) -> Iterator[Pat
+    )
+
+    # extra commit
+    (tmp_path / "foo").write_text("foo", encoding="utf-8")
```
Is this an empty commit because it writes the same as the first commit? (It probably does not matter, I am just wondering.)

mikhainin replied:
Thanks for flagging this, I added a new file instead

radoering commented on tests/vcs/git/test_backend.py:
```diff
@@ -75,3 +82,31 @@ def test_urlpathjoin(url: str, expected_result: str) -> None:
     assert result == expected_result
+
+
+@pytest.mark.skipif(GIT_NOT_INSTALLLED, reason="These tests requires git cli")
```
Do they? I thought they just use `dulwich` and do not need git installed (in comparison to `SystemGit`)?

mikhainin replied:
I thought the fixture uses `git cli` but now I see it's wasn't the case. Removed this `skipif`

radoering commented on tests/vcs/git/test_backend.py:
```diff
@@ -75,3 +82,30 @@ def test_urlpathjoin(url: str, expected_result: str) -> None:
     assert result == expected_result
+
+
+class TestBackendGit:
```
I think we do not need this class anymore (without the `skipif`). There are already other tests for the `Git` class, which are top level. (In general, we tend to not create test classes but just test functions.)

mikhainin replied:
removed the class

radoering commented on tests/vcs/git/git_fixture.py:
```diff
@@ -0,0 +1,21 @@
+    import dulwich.repo
+
+
+GIT_NOT_INSTALLLED = shutil.which("git") is None
```
I think we can put this constant back into `test_system` because it is unlikely that it is relevant for other tests than for `SystemGit`.

mikhainin replied:
done

radoering commented on tests/conftest.py:
```diff
@@ -578,3 +584,49 @@ def project_context(project: str | Path, in_place: bool = False) -> Iterator[Pat
+
+
+@pytest.fixture()
+def temp_repo(tmp_path: Path) -> TempRepoFixture:
```
This fixture is only used in `test_backend` and `test_system`, isn't it? Then, we can put it in a `conftest.py` in `tests/vcs/git`.

mikhainin replied:
Yes. However, it will interfere with Ruff's rules as ruff will be removed as an "unused import".
https://github.com/astral-sh/ruff/issues/3295
https://github.com/astral-sh/ruff/issues/8796

We can mark the import with `noqa` but I wasn't sure it was the best approach

radoering replied:
We do not have to import a fixture from a `conftest.py` to use it or do you mean something else?

mikhainin replied:
ah, sure, moved to `tests/vcs/git/conftest.py`

jelmer commented on src/poetry/vcs/git/backend.py:
```diff
@@ -313,21 +314,38 @@ def _clone(cls, url: str, refspec: GitRefSpec, target: Path) -> Repo:
 
+    @classmethod
+    def _find_object_by_sha_prefix(cls, repo: Repo, sha_prefix: bytes) -> Commit | None:
+        for sha in repo.object_store:
```
```suggestion
        for sha in repo.object_store.iter_prefix(prefix):
```

abn commented on src/poetry/vcs/git/backend.py:
```diff
@@ -313,21 +314,38 @@ def _clone(cls, url: str, refspec: GitRefSpec, target: Path) -> Repo:
-                f"Failed to clone {url} at '{refspec.key}', verify ref exists on"
-                " remote."
-            )
+            if refspec.is_sha_short:
```
@mikhainin can we move the new logic into an `else` block under the for loop here;

https://github.com/python-poetry/poetry/blob/6f6fd7012983a2e749c6030c1f5f155fd4397058/src/poetry/vcs/git/backend.py#L95-L101

And as @jelmer suggested, we can simplify it to something like 

```py
            else:
                for sha in repo.object_store.iter_prefix(prefix):
                    self.revision = sha.decode("utf-8")
                    break
```

Assuming we get a dulwich release with the fix soon.

jelmer replied:
I'll see if I can do a release this weekend,

jelmer replied:
0.22.6 is now out with the fix

dimbleby replied:
No it isn't!  I guess your release pipeline failed?

mikhainin replied:
Right, I moved the logic to `_normalise()` - it now looks even cleaner.
Waiting for **0.22.6** to appear and then will push the change.

Secrus replied:
Dulwich 0.22.6 is out

abn commented:
This will likely also require #9849 if we intend to update dulwich.

dimbleby commented:
> This will likely also require #9849 if we intend to update dulwich.

Since recent dulwich releases are bumping only the patch number, poetry users are already getting the latest dulwich.

Are you saying that dulwich has released a breaking change in a patch release?

jelmer commented:
I'm curious what change is, we haven't had reports of API breakages.

abn commented:
@jelmer To close the loop regarding @dimbleby 's question; I have raised a dulwich issue for triage with a reproducer.

Cross-post: https://github.com/jelmer/dulwich/issues/1443
